azurerm/internal/services/containers/kubernetes_cluster_node_pool_resource.go

```diff
@@ -707,3 +720,43 @@ func resourceKubernetesClusterNodePoolDelete(d *schema.ResourceData, meta interf
 
 	return nil
 }
+
+func upgradeSettingsSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"max_surge": {
+					Type:     schema.TypeString,
+					Optional: true,
+				},
+			},
+		},
+	}
+}
+
+func expandUpgradeSettings(input []interface{}) *containerservice.AgentPoolUpgradeSettings {
+	if len(input) == 0 {
+		return nil
+	}
+	upgradeSettingInput := input[0].(map[string]interface{})
+	upgradeSetting := containerservice.AgentPoolUpgradeSettings{}
+
+	if maxSurgeRaw := upgradeSettingInput["max_surge"].(string); maxSurgeRaw != "" {
+		upgradeSetting.MaxSurge = utils.String(maxSurgeRaw)
+	}
+	return &upgradeSetting
+}
+
+func flattenUpgradeSettings(input *containerservice.AgentPoolUpgradeSettings) []interface{} {
+	upgradeSettings := make([]interface{}, 0)
+
+	if input == nil {
+		return upgradeSettings
+	}
+	nodePoolSetting := make(map[string]interface{})
+	nodePoolSetting["max_surge"] = *input.MaxSurge
+	return append(upgradeSettings, nodePoolSetting)
```

Comment:
we need to ensure this field always has a value if the block isn't present - that is, it's possible for the Azure API to return this block without it necessarily having values for the keys within it - to help identify that we should make this:

```suggestion
	if input == nil {
		return []interface{}{}
	}

	maxSurge := ""
	if input.MaxSurge != nil {
		maxSurge = *input.MaxSurge
	}

	return []interface{}{
		map[string]interface{}{
			"max_surge": maxSurge,
		},
	}
```